1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,232 @@
+= API Contracts
+
+The "I" means interface, and so writing a contract for that interface is the act of writing down what that interface is going to look like, to avoid folks having to guess, or go and hunt for some documentation somewhere.
+
+Let's use the fantastic https://pokeapi.co/[PokeAPI] as an example.
+
+[source,javascript]
+----
+{
+  "id": 12,
+  "name": "butterfree",
+  "base_experience": 178,
+  "height": 11,
+  "is_default": true,
+  "order": 16,
+  "weight": 320,
+  "abilities": [{
+    "is_hidden": true,
+    "slot": 3
+  }]
+}
+----
+
+If we look just at the data, we can deduce that sure, ID is probably an auto-incrementing identifier, the name seems to be a lower-cased English string, but after that it starts getting a bit odd.
+
+Height and weight, I guess there is no imperial and metric in the game just units, but what is order all about? The order that pokemon is sat in my current party? That Ash bumped into in the cartoon? In order of cuteness? What?!
+
+An API contract can help answer most of the questions and more for anyone wanting to understand how an API is meant to work.
+
+- Which resources or methods are available and how do you interact with them?
+- Which HTTP endpoints are required for that endpoint, for things like authentication?
```

Comment:
Do you mean "Which HTTP endpoints are related to that endpoint, for things like authentication?" (eg. auth to  token) or "Which HTTP endpoints require authentication"? Or something else?

Reply:
Ha, "Which HTTP **headers** are required for that endpoint, for things like authentication?" :)